### PR TITLE
feat: add extensions slot to KnowledgeUnit schema

### PR DESCRIPTION
**Pull request by peteski22**
## Summary

- Add an optional `extensions` field to `KnowledgeUnit` as the single escape hatch from `additionalProperties: false`, allowing implementation-specific fields with namespaced keys (`namespace:key` format)
- Enforce key format (`^[a-z0-9][a-z0-9_-]*:\S+$`) in all SDK layers (Go, Python, CLI handler) so consumers without the server get the same guarantees
- Server relies on SDK validation rather than duplicating it; service layer catches SDK `ValidationError` and maps to 422

## Test plan

- [x] Table-driven extension key validation tests in Go SDK (valid keys, invalid keys including whitespace edge cases)
- [x] Parameterized extension key validation tests in Python SDK
- [x] Schema-as-oracle tests validate serialized models against canonical JSON Schema
- [x] CLI MCP handler test for bad key rejection
- [x] Server integration tests for propose with/without extensions and bad key rejection
- [x] Schema-level tests for fixture validation and bad key rejection via `patternProperties`
- [x] `make lint` and `make test` pass (server `ty` requires local SDK install during development)

Closes #411

## Release Notes

* **New Features**
  * Added optional `extensions` support on knowledge units, validated using `namespace:key` keys
  * Extensions can now be provided when proposing knowledge units via the server API, CLI, and SDKs

* **Bug Fixes**
  * Improved request handling to validate `extensions` and return clear `422` errors for invalid formats

* **Tests**
  * Expanded coverage across Go, Python, and backend API tests, including schema-oracle checks

* **Chores**
  * Updated CLI module resolution to use the local SDK copy during builds

### CODE REVIEW

File: cli/go.mod [hunk 1]
```diff
@@ -37,4 +37,4 @@ require (
 
 // Monorepo: the SDK is consumed locally. Re-pin to a published version
 // when the SDK is released alongside the CLI.
-//replace github.com/mozilla-ai/cq/sdk/go => ../sdk/go
+replace github.com/mozilla-ai/cq/sdk/go => ../sdk/go
```
[the remark this answers is not in the document]
Comment (peteski22) [reply cli/go.mod:40]: Intentional for local development; a follow-up PR reverts it once the SDK is published to the module proxy.

File: cli/mcpserver/propose_test.go [hunk 2]
```diff
@@ -181,6 +211,29 @@ func TestHandlePropose(t *testing.T) {
...
+						"bad-key": "value",
+					},
+				},
+			},
+		})
+		require.NoError(t, err)
+		require.True(t, result.IsError)
+		require.Contains(t, result.Content[0].(mcp.TextContent).Text, "namespace:key")
+	})
+
```
[the remark this answers is not in the document]
Comment (peteski22) [reply cli/mcpserver/propose_test.go:236]: Fixed. Added `"rejects non-object extensions argument"` subtest.

File: cli/mcpserver/propose.go [hunk 3]
```diff
@@ -66,6 +73,17 @@ func (s *Server) HandlePropose(ctx context.Context, req mcp.CallToolRequest) (*m
 		return mcp.NewToolResultError("domains must contain at least one tag"), nil
 	}
 
+	var extensions map[string]any
+	if raw, ok := req.GetArguments()["extensions"]; ok {
+		if m, ok := raw.(map[string]any); ok {
+			extensions = m
+		}
+	}
```
[the remark this answers is not in the document]
Comment (peteski22) [reply cli/mcpserver/propose.go:83]: Fixed. Non-map values now return a tool error instead of being silently dropped.

File: schema/python/tests/test_cq_schema.py [hunk 4]
```diff
@@ -69,6 +69,52 @@ def test_scoring_values_validates_against_scoring_schema() -> None:
...
+        "id": "ku_00000000000000000000000000000099",
+        "domains": ["test"],
+        "insight": {"summary": "s", "detail": "d", "action": "a"},
+        "extensions": {
+            "impl:key": "string-value",
+            "my-impl:nested": {"a": 1},
+            "x0:flag": True,
+        },
+    }
+    jsonschema.validate(instance=instance, schema=schema)
```
[the remark this answers is not in the document]
Comment (peteski22) [reply schema/python/tests/test_cq_schema.py:132]: Fixed. Added `test_extensions_reject_more_than_twenty_keys` to lock in the schema constraint.

File: sdk/go/extensions.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+var extensionKeyPattern = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]*:\S+$`)
+
+// ValidateExtensionKeys checks that every key in extensions uses namespace:key format.
+func ValidateExtensionKeys(extensions map[string]any) error {
+	for k := range extensions {
+		if !extensionKeyPattern.MatchString(k) {
+			return fmt.Errorf("extension key %s must match namespace:key format (e.g. myimpl:field)", k)
+		}
+	}
+	return nil
```
[the remark this answers is not in the document]
Comment (peteski22) [reply sdk/go/extensions.go:18]: `maxProperties: 20` is intentionally a soft cap enforced only at the schema level, per the RFC discussion. Runtime enforcement is not applied so implementations can exceed the cap when needed; the schema constraint serves as a guideline for interoperability.

File: sdk/python/src/cq/models.py [hunk 6]
```diff
@@ -102,6 +104,7 @@ class KnowledgeUnit(BaseModel):
     tier: Tier = Tier.LOCAL
     created_by: str = ""
     superseded_by: str | None = None
+    extensions: dict[str, Any] | None = None
```
[the remark this answers is not in the document]
Comment (peteski22) [reply sdk/python/src/cq/models.py:107]: Same as above; `maxProperties: 20` is a soft schema-only cap per the RFC. No runtime enforcement by design.

File: sdk/go/extensions.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+}
```
[the remark this answers is not in the document]
Comment (peteski22) [reply sdk/go/extensions.go:19]: `maxProperties: 20` is intentionally a soft cap enforced only at the schema level, per the RFC discussion. Runtime enforcement is not applied so implementations can exceed the cap when needed; the schema constraint serves as a guideline for interoperability.

File: sdk/python/src/cq/models.py [hunk 8]
```diff
@@ -4,13 +4,15 @@
...
 from enum import StrEnum
+from typing import Any
 
 from pydantic import BaseModel, Field, field_validator, model_validator
 
 from ._util import _as_list
 
 _KU_ID_PREFIX = "ku_"
 _KU_ID_PATTERN = re.compile(r"^ku_[0-9a-f]{32}$")
+_EXTENSION_KEY_PATTERN = re.compile(r"^[a-z0-9][a-z0-9_-]*:\S+$")
```
[the remark this answers is not in the document]
Comment (peteski22) [reply sdk/python/src/cq/models.py:17]: Same as above; `maxProperties: 20` is a soft schema-only cap per the RFC. No runtime enforcement by design.

File: sdk/python/src/cq/models.py [hunk 9]
```diff
@@ -120,6 +123,17 @@ def _validate_superseded_by_format(cls, v: str | None) -> str | None:
...
+    @field_validator("extensions")
+    @classmethod
+    def _validate_extension_keys(cls, v: dict[str, Any] | None) -> dict[str, Any] | None:
+        """Enforce namespace:key format on extension keys."""
+        if v is None:
+            return v
+        bad = [k for k in v if not _EXTENSION_KEY_PATTERN.match(k)]
+        if bad:
+            raise ValueError(f"extension keys must match namespace:key format, got: {bad}")
+        return v
```
[the remark this answers is not in the document]
Comment (peteski22) [reply sdk/python/src/cq/models.py:135]: Same as above; `maxProperties: 20` is a soft schema-only cap per the RFC. No runtime enforcement by design.
